src/Nethermind/Nethermind.Evm/Tracing/GethStyle/Custom/Native/Call/NativeCallTracer.cs

```diff
@@ -0,0 +1,262 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
+using Nethermind.Core;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Int256;
+using Nethermind.Serialization.Json;
+
+namespace Nethermind.Evm.Tracing.GethStyle.Custom.Native.Call;
+
+// The callTracer tracks all the call frames executed during a transaction, including depth 0.
+// The result will be a nested list of call frames, resembling how the EVM works.
+// They form a tree with the top-level call at root and sub-calls as children of the higher levels.
+//
+// TracerConfig options:
+// onlyTopCall (default = false): Only the main (top-level) call will be processed to avoid any extra processing if only the main call info is required.
+// withLog (default = false): Logs emitted during each call will also be collected and included in the result.
+public sealed class NativeCallTracer : GethLikeNativeTxTracer
+{
+    public const string CallTracer = "callTracer";
+
+    private readonly long _gasLimit;
+    private readonly NativeCallTracerConfig _config;
+    private readonly List<NativeCallTracerCallFrame> _callStack = [];
+
+    private NativeCallTracerCallFrame? _firstCallFrame;
+    private EvmExceptionType? _error;
+    private long _remainingGas;
+
+
+    public NativeCallTracer(
+        Transaction? tx,
+        GethTraceOptions options) : base(options)
+    {
+        IsTracingActions = true;
+        _gasLimit = tx!.GasLimit;
+
+        _config = options.TracerConfig?.Deserialize<NativeCallTracerConfig>(EthereumJsonSerializer.JsonOptions) ?? new NativeCallTracerConfig();
+
+        if (_config.WithLog)
+        {
+            IsTracingLogs = true;
+        }
+    }
+
+    protected override GethLikeTxTrace CreateTrace() => new(_firstCallFrame);
+
+    public override GethLikeTxTrace BuildResult()
+    {
+        GethLikeTxTrace result = base.BuildResult();
+        _firstCallFrame = _callStack[0];
+        result.CustomTracerResult = new GethLikeCustomTrace { Value = _firstCallFrame };
+
+        for (int i = 1; i < _callStack.Count; i++)
+        {
+            _callStack[i].Dispose();
+        }
```

Comment:
I am a bit confused, we have a call stack that we dispose here and we dispose `NativeCallTracerCallFrame childCallFrame in Calls` in `NativeCallTracerCallFrame`, aren't we disposing same objects? What is the difference?

Reply:
At this point in the code, `_callStack` should only contain one value, the top call frame. The first call frame is what's then passed into the trace and serialized into JSON. I added this loop as a protection against any potential memory leaks in case an unforseen error in the tracer causes the callStack to incorretly nest child elements and the elements are left in the top level.

However, this code wouldn't be necessary if I implement your suggestion above about making the call stack be an ArrayPoolList.

Reply:
Paired and resolved synchronously